<a href="https://colab.research.google.com/github/poudyaldiksha/Data-Science-project/blob/main/Lesson_20_b2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 20: Time Series Data

---


We will  analyse data on air quality in one of the cities in Italy. We need to determine whether the concentration of air pollutants increase or decrease over time.

In the process, we will also learn how to work with time-series data, i.e, how to work with date and time values.


---

### Data Description



|Index|Columns|Description|
|-|-|-|
|0|Date|Date (DD/MM/YYYY)|
|1|Time|Time (HH.MM.SS)|
|2|CO(GT)|True hourly averaged concentration $\text{CO}$ in $\frac{mg}{m^3}$ (reference analyzer)|
|3|PT08.S1(CO)|PT08.S1 (tin oxide) hourly averaged sensor response (nominally $\text{CO}$ targeted)|
|4|NMHC(GT)|True hourly averaged overall Non-Methane Hydrocarbons concentration in $\frac{\mu g}{m^3}$|
|5|C6H6(GT)|True hourly averaged Benzene concentration in $\frac{\mu g}{m^3}$||
|6|PT08.S2(NMHC)|PT08.S2 (titania) hourly averaged sensor response (nominally $\text{NMHC}$ targeted)|
|7|NOx(GT)|True hourly averaged $\text{NO}_x$ concentration in Parts per billion (ppb)|
|8|PT08.S3(NOx)|PT08.S3 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_x$ targeted)|
|9|NO2(GT) |True hourly averaged $\text{NO}_2$ concentration in $\frac{\mu g}{m^3}$|
|10|PT08.S4(NO2)|PT08.S4 (tungsten oxide) hourly averaged sensor response (nominally $\text{NO}_2$ targeted)|
|11|PT08.S5(O3) |PT08.S5 (indium oxide) hourly averaged sensor response (nominally $\text{O}_3$ targeted)|
|12|T|Temperature in °C|
|13|RH|Relative Humidity (%)|
|14|AH|AH Absolute Humidity|



**Source:** https://archive.ics.uci.edu/ml/datasets/Air+quality

---

---

### Activity 1: Loading Data Sets


The values in the dataset are separated by semi-columns (`;`) instead of commas.

To read the given CSV file, pass the `sep = ';'` parameter inside the `read_csv()` function along with the file location.

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
s = 12,3
type(s)

tuple

In [ ]:
s = 12,
type(s)

tuple

In [ ]:
# Import the required modules and load the time-series dataset on air quality
import pandas as pd

air_df=pd.read_csv("/content/drive/MyDrive/datasets/AirQualityUCI.csv", sep=";")
air_df.head()


Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1  10/03/2004  19.00.00      2       1292.0     112.0      9,4          955.0   
2  10/03/2004  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3  10/03/2004  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4  10/03/2004  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

In [ ]:
air_df.dtypes

Date              object
Time              object
CO(GT)            object
PT08.S1(CO)      float64
NMHC(GT)         float64
C6H6(GT)          object
PT08.S2(NMHC)    float64
NOx(GT)          float64
PT08.S3(NOx)     float64
NO2(GT)          float64
PT08.S4(NO2)     float64
PT08.S5(O3)      float64
T                 object
RH                object
AH                object
Unnamed: 15      float64
Unnamed: 16      float64
dtype: object

In [ ]:
air_df.head()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1  10/03/2004  19.00.00      2       1292.0     112.0      9,4          955.0   
2  10/03/2004  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3  10/03/2004  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4  10/03/2004  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

In [ ]:
air_df.tail(20)

Date Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
9451  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9452  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9453  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9454  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9455  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9456  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9457  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9458  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9459  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9460  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9461  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9462  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9463  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9464  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9465  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9466  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9467  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9468  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9469  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9470  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   

      PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)    T   RH   AH  \
9451           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9452           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9453           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9454           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9455           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9456           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9457           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9458           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9459           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9460           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9461           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9462           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9463           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9464           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9465           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9466           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9467           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9468           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9469           NaN      NaN           NaN          NaN  NaN  NaN  NaN   
9470           NaN      NaN           NaN          NaN  NaN  NaN  NaN   

      Unnamed: 15  Unnamed: 16  
9451          NaN          NaN  
9452          NaN          NaN  
9453          NaN          NaN  
9454          NaN          NaN  
9455          NaN          NaN  
9456          NaN          NaN  
9457          NaN          NaN  
9458          NaN          NaN  
9459          NaN          NaN  
9460          NaN          NaN  
9461          NaN          NaN  
9462          NaN          NaN  
9463          NaN          NaN  
9464          NaN          NaN  
9465          NaN          NaN  
9466          NaN          NaN  
9467          NaN          NaN  
9468          NaN          NaN  
9469          NaN          NaN  
9470          NaN          NaN

In [ ]:
# Get information about dataframe
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   object 
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   object 
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   object 
 13  RH             9357 non-null   object 
 14  AH             9357 non-null   object 
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(10), object(7)
memory usage: 1.2+ MB


---

### Activity 2: Missing Values



In [ ]:
# Check for the missing values in the  DataFrame.
air_df.isnull().sum()

Date              114
Time              114
CO(GT)            114
PT08.S1(CO)       114
NMHC(GT)          114
C6H6(GT)          114
PT08.S2(NMHC)     114
NOx(GT)           114
PT08.S3(NOx)      114
NO2(GT)           114
PT08.S4(NO2)      114
PT08.S5(O3)       114
T                 114
RH                114
AH                114
Unnamed: 15      9471
Unnamed: 16      9471
dtype: int64

In [ ]:
#Drop the 'Unnamed: 15 and Unnamed: 16' columns from the DataFrame.

air_df.drop(columns=["Unnamed: 15","Unnamed: 16"],inplace = True)

In [ ]:
# Activity:
air_df

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  \
0     10/03/2004  18.00.00    2,6       1360.0     150.0     11,9   
1     10/03/2004  19.00.00      2       1292.0     112.0      9,4   
2     10/03/2004  20.00.00    2,2       1402.0      88.0      9,0   
3     10/03/2004  21.00.00    2,2       1376.0      80.0      9,2   
4     10/03/2004  22.00.00    1,6       1272.0      51.0      6,5   
...          ...       ...    ...          ...       ...      ...   
9466         NaN       NaN    NaN          NaN       NaN      NaN   
9467         NaN       NaN    NaN          NaN       NaN      NaN   
9468         NaN       NaN    NaN          NaN       NaN      NaN   
9469         NaN       NaN    NaN          NaN       NaN      NaN   
9470         NaN       NaN    NaN          NaN       NaN      NaN   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
0            1046.0    166.0        1056.0    113.0        1692.0   
1             955.0    103.0        1174.0     92.0        1559.0   
2             939.0    131.0        1140.0    114.0        1555.0   
3             948.0    172.0        1092.0    122.0        1584.0   
4             836.0    131.0        1205.0    116.0        1490.0   
...             ...      ...           ...      ...           ...   
9466            NaN      NaN           NaN      NaN           NaN   
9467            NaN      NaN           NaN      NaN           NaN   
9468            NaN      NaN           NaN      NaN           NaN   
9469            NaN      NaN           NaN      NaN           NaN   
9470            NaN      NaN           NaN      NaN           NaN   

      PT08.S5(O3)     T    RH      AH  
0          1268.0  13,6  48,9  0,7578  
1           972.0  13,3  47,7  0,7255  
2          1074.0  11,9  54,0  0,7502  
3          1203.0  11,0  60,0  0,7867  
4          1110.0  11,2  59,6  0,7888  
...           ...   ...   ...     ...  
9466          NaN   NaN   NaN     NaN  
9467          NaN   NaN   NaN     NaN  
9468          NaN   NaN   NaN     NaN  
9469          NaN   NaN   NaN     NaN  
9470          NaN   NaN   NaN     NaN  

[9471 rows x 15 columns]

We know that the other columns have 114 missing values. Let's see which rows in the `Date` column are missing.

In [ ]:
a = air_df["Date"].isnull()


In [ ]:
# Print the rows missing in the 'Date' column.
air_df[air_df["Date"].isnull()== True]

Date Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
9357  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9358  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9359  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9360  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9361  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
...   ...  ...    ...          ...       ...      ...            ...      ...   
9466  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9467  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9468  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9469  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9470  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   

      PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)    T   RH   AH  
9357           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9358           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9359           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9360           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9361           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
...            ...      ...           ...          ...  ...  ...  ...  
9466           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9467           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9468           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9469           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9470           NaN      NaN           NaN          NaN  NaN  NaN  NaN  

[114 rows x 15 columns]

**The `set_option()` Function**

This method allows us to configure the display to show a complete data frame instead of a truncated one. A function set_option() is provided by pandas to display all rows of the data frame. display.max_rows represents the maximum number of rows that pandas will display while displaying a data frame.

The default value of max_rows is 10. If set to ‘None‘ then it means all rows of the data frame. Set value of display.max_rows to None and pass it to set_option and this will display all rows from the data frame.



In [ ]:
#pd.set_option("display.max_rows",None)
#df

We can see that all the rows starting from index `9357` to `9470` contain missing values in all the columns.

**The `dropna()` Function**

Let's drop them all using the `dropna()` function. It removes the rows or columns containing the missing values. If no parameter is passed to the `dropna()` function, then it removes all the rows containing at least one null value.

In [ ]:
#Drop the rows containing at least one null value in the  DataFrame using the 'dropna()' function.
air_df.dropna(inplace = True)

Now, let's check whether all the missing values are removed or not.

In [ ]:
#  Check whether all the missing values are removed or not.

air_df.isnull().sum()

Date             0
Time             0
CO(GT)           0
PT08.S1(CO)      0
NMHC(GT)         0
C6H6(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               0
dtype: int64

In [ ]:
#Activity : get number of rows and columns
air_df.shape

(9357, 15)

In [ ]:
# Print the last 5 rows of the DataFrame.

air_df.tail()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  \
9352  04/04/2005  10.00.00    3,1       1314.0    -200.0     13,5   
9353  04/04/2005  11.00.00    2,4       1163.0    -200.0     11,4   
9354  04/04/2005  12.00.00    2,4       1142.0    -200.0     12,4   
9355  04/04/2005  13.00.00    2,1       1003.0    -200.0      9,5   
9356  04/04/2005  14.00.00    2,2       1071.0    -200.0     11,9   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
9352         1101.0    472.0         539.0    190.0        1374.0   
9353         1027.0    353.0         604.0    179.0        1264.0   
9354         1063.0    293.0         603.0    175.0        1241.0   
9355          961.0    235.0         702.0    156.0        1041.0   
9356         1047.0    265.0         654.0    168.0        1129.0   

      PT08.S5(O3)     T    RH      AH  
9352       1729.0  21,9  29,3  0,7568  
9353       1269.0  24,3  23,7  0,7119  
9354       1092.0  26,9  18,3  0,6406  
9355        770.0  28,3  13,5  0,5139  
9356        816.0  28,5  13,1  0,5028

In [ ]:
"sunshine.....".replace(".",":")

'sunshine:::::'

In [ ]:
time = []
for i in air_df["Time"]:
   time.append(i.replace(".",":"))


In [ ]:
type(time)

list

In [ ]:
time[:10]

['18:00:00',
 '19:00:00',
 '20:00:00',
 '21:00:00',
 '22:00:00',
 '23:00:00',
 '00:00:00',
 '01:00:00',
 '02:00:00',
 '03:00:00']

In [ ]:
time_s = pd.Series(time)
time_s

0       18:00:00
1       19:00:00
2       20:00:00
3       21:00:00
4       22:00:00
          ...   
9352    10:00:00
9353    11:00:00
9354    12:00:00
9355    13:00:00
9356    14:00:00
Length: 9357, dtype: object

In [ ]:
a = pd.Series([1,2,3])
b= pd.Series([4,5,6])
a+b

0    5
1    7
2    9
dtype: int64

0       10/03/2004 18:00:00
1       10/03/2004 19:00:00
2       10/03/2004 20:00:00
3       10/03/2004 21:00:00
4       10/03/2004 22:00:00
               ...         
9352    04/04/2005 10:00:00
9353    04/04/2005 11:00:00
9354    04/04/2005 12:00:00
9355    04/04/2005 13:00:00
9356    04/04/2005 14:00:00
Length: 9357, dtype: object

In [ ]:
type(dt_series)

pandas.core.series.Series

In [ ]:
a = pd.Series(["a","b","c"])
b= pd.Series(["e","f","g"])
a+b

0    ae
1    bf
2    cg
dtype: object

---

### Activity 3: Converting `datetime` Values - `pd.to_datetime()` Function

 A `datetime` value is a combination of date and time.



**Syntax:** `pd.to_datetime(series)`

where the `series` is the Pandas series whose values are to be converted into `datetime` values.

The default format of a `datetime` value is `YYYY-MM-DD` where `YYYY` denotes year, `MM` denotes month and `DD` denotes day.

But first we will have to replace the periods (or dots) with colons (`:`) in the values of the `Time` column because the time values are separated colons, i.e., `HH:MM:SS` where

- `HH` denotes hours

- `MM` denotes minutes

- `SS` denotes seconds

In [ ]:
#Concatenate the values stored in the 'Date' and 'Time' columns

dt_series = air_df["Date"] + " " + time_s
dt_series

In [ ]:
date_time = pd.to_datetime(dt_series, format="%d/%m/%Y %H:%M:%S")
date_time

0      2004-03-10 18:00:00
1      2004-03-10 19:00:00
2      2004-03-10 20:00:00
3      2004-03-10 21:00:00
4      2004-03-10 22:00:00
               ...        
9352   2005-04-04 10:00:00
9353   2005-04-04 11:00:00
9354   2005-04-04 12:00:00
9355   2005-04-04 13:00:00
9356   2005-04-04 14:00:00
Length: 9357, dtype: datetime64[ns]

---

### Activity 4: Parsing `datetime` Values - `date(), strftime()` & `time()` Functions

Let's learn how to extract date, day, month, year, time, hours, minutes and seconds from a `datetime` value.

**The `date()` Function**

In general, a `datetime` value is a `Timestamp` object. It holds both date and time.

In [ ]:
# Get the first 'datetime' value
timestamp_0 = date_time[0]
timestamp_0

Timestamp('2004-03-10 18:00:00')

To get a date (i.e., `datetime.date` object) from a timestamp, apply the `date()` function.

In [ ]:
# Get date from the timestamp stored in the 'timestamp_0' variable
# Print the value and its data-type.
date_0=timestamp_0.date()
date_0

datetime.date(2004, 3, 10)

**The `day, month` & `year` Attributes**

To get day, month and year from a `datetime.date` value apply the `day, month` and `year` attributes.

The output will be day, month and year as integer values.

In [ ]:
# Get the day, month and year attributes from the date stored in the 'date_0' variable
print(date_0.year)
print(date_0.month)
print(date_0.day)

2004
3
10


**The `strftime()` Function**

To convert a `datetime.date` value into a string value, use the `strftime()` function. You can also format the date in different formats using the `strftime()` function.

Let's format the date stored in the `date_0` variable in the `10-Mar-2004` format. To do this, you will have to pass the following input to the `strftime()` function.

`'%d-%b-%Y'`

It is a combination of three format codes. They are:

1. `%d` format code denotes day of the month as a zero-padded decimal number. E.g., 01, 02, 03 etc.

2. `%b` format code denotes abbreviated month name of month in the `datetime.date` value. E.g., Jan Feb, March etc.

3. `%Y` format code denotes year with century as a decimal number. E.g, 2019, 2020, 2021 etc.

The hyphen `-` used here just to separate the day, month and year values. You can separate them with different characters.

To get a list of all the `strftime()` format codes, click on the link provided below and then go to the **`strftime()` and `strptime() Format Codes`** section.

https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes



In [ ]:
# Format the date stored in the 'date_0' variable in the '10-Mar-2004' format.
print(date_0.strftime('%D-%B-%Y'))

03/10/04-March-2004


In [ ]:
# Format the date stored in the 'date_0' variable in the 'March 10, 2004' format.
print(date_0.strftime('%B %d,%Y'))

March 10,2004


**The `time()` Function**

To get time from a timestamp object, apply the `time()` function. It returns a `datetime.time` object.

In [ ]:
# Get the time from the timestamp stored in the 'timestamp_0' variable.
time_0=timestamp_0.time()
time_0

**The `hour, minute` & `second` Attributes**

To get hour, minute and second values from a `datetime.time` object, apply the `hour, minute` and `second` attributes.

The output will be hour, minute and second as integer values.

In [ ]:
#Get the hour, minute and second attributes from the time stored in the 'time_0' variable.
print(time_0.hour)
print(time_0.minute)
print(time_0.second)


---

# datetime Module:

- **datetime Module:** This is a general-purpose module in the Python standard library for working with dates and times. It provides classes like date, time, datetime, and timedelta to represent dates, times, timedeltas (durations), and perform basic operations like creating, formatting, and manipulating date-time objects. It's suitable for general date-time handling in Python, not just for pandas DataFrames.

- **pd.to_datetime:** This function is specifically part of the pandas library. It's used to convert strings, integers, timestamps, or other objects into pandas-specific DatetimeIndex objects. These objects are optimized for efficient time series data analysis within pandas DataFrames. They allow for advanced functionalities like indexing, resampling, date manipulation, and working with time zones, which are essential for pandas data analysis.

**Key points:**
- Use the datetime module for general date-time needs in Python, outside of pandas.
- Use pd.to_datetime when working with time series data within pandas DataFrames to leverage their optimized functionality and integration with other pandas tools.

### Activity 5: The `datetime.now()` function

The `datetime` module allows us to create date and time objects and manipulate them as we desire. Let's start with importing the `datetime` module.

In [ ]:
# Import the 'datetime' module.
import datetime

Let's now print the current date and time using the `datetime.now()`. It returns the current time in Greenwich Mean Time (GMT) or Coordinated Universal Time (UTC) time zone.

The syntax for using the `datetime.now()` function is `datetime.datetime.now()` where the first occurrence of `datetime` denotes the module name and its second occurrence denotes the object type.

**Note:** The default format of the `datetime` object is `YYYY-MM-DD HH:MM:SS` where

- `YYYY` denotes year,

- `MM` denotes month,

- `DD` denotes day,

- `HH` denotes hours in 24-hours,

- `MM` denotes minutes, and

- `SS` denotes seconds

In [ ]:
# Get the current time, print it and its data-type.
time_now = datetime.datetime.now()
print(time_now)

The fractional value in the above time is microseconds. That's the level of precision of a `datetime` object.

You can apply the `day, month, year, hour, minute` and `second` attributes to get the corresponding day, month, year, hour, minute and second values.

In [ ]:
# Get the corresponding day, month, year, hour, minute and second values.
print(time_now.day)
print(time_now.month)
print(time_now.hour)
print(time_now.minute)
print(time_now.second)

---

### Activity 6: The `timedelta` Object

To get time in Indian Standard Time (IST) time zone, we have to add 5 hours and 30 minutes to the UTC time because IST is equivalent to UTC+5:30. For this, we need to use the `timedelta` object. It represents a duration, the difference between two dates or times.

Inside the `timedelta` object, you may pass any one of the following parameters (or arguments).

- `days`,

- `seconds=0`,

- `microseconds`,

- `milliseconds`,

- `minutes`,

- `hours`, and

- `weeks`

The default value of each of the above parameters is zero. Let's pass `hours=5` and `minutes=30` parameters to the `timedelta` object. Then add it to the current time.


In [ ]:
#Get the current time in IST (or UTC+5:30) time zone. Print it and its data-type.
time_ist = time_now + datetime.timedelta(hours = 5,minutes = 30)
print(time_ist)

You can also convert the datetime in any time zone by simply adding or subtracting the hours and minutes values from the current time. As a practice, let's convert the current time in the UTC-7:00 zone. It's the time zone for the pacific region of the USA. The silicon valley is located in the same regions.

In [ ]:
# Get the current time in the UTC-7 time zone.
time_cal = time_now + datetime.timedelta(hours = 1)
print(time_cal)

---

### Activity 7: The `datetime.strftime()` Function

The current date and time values are `datetime` objects. Let's convert them into a string value using the `strftime()` function in the `Month Day, Year HH:MM:SS AM/PM` format where `HH` hours in 12-hour format.

Click on the link provided below to get the format codes.

[The `strftime()` & `strptime()` Format Codes](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior)

**Note:** You may choose any other format of your liking to format a date and time value.

In [ ]:
# Convert the current date and time into a string. Also print the data-type of the final value.
a = datetime.datetime.now() + datetime.timedelta(minutes=30, hours=5)
a = a.strftime('%B %z, %Y %I:%M:%S %p')
a

In [ ]:
curr_time = datetime.datetime.now()
curr_time = curr_time.strftime('%m %d, %y %M:%S:%I')
curr_time

In [ ]:
a = datetime.datetime.now() + datetime.timedelta(minutes=30, hours=5)
a.strftime('%a %d %b %Y, %I:%M%p')

In [ ]:
type(curr_time)

---

### Activity 8: The `date()` & `time()` Functions

Let's first separate the date and time values from the current `datetime` object using the `date()` and `time()` functions respectively.

Also, let's convert the date and time values into string and format them in the `Month Day, Year` and `HH:MM:SS AM/PM` date and time formats respectively.

In [ ]:
# Separate the date and time values from the 'datetime' object. Convert them into string
b = datetime.datetime.now() + datetime.timedelta(minutes=30, hours=5)
b_date = b.date().strftime('%B %d, %Y')
b_time = b.time().strftime('%I:%M:%S %p')
print(b_date,b_time)

---

### Activity 9: The `datetime.strptime()` Function

We can create a `datetime` object by passing the string values to the `strptime()` function.

As an example, let's create a `datetime` object for the date December 26, 2024.


The `strptime()` function requires two parameters:  

- The `datetime` object as a string

- The combination of format codes for the passed `datetime` object. E.g., if the datetime object is `May 04, 2020 06:14:48 PM`, then the required combination format codes is `'%B %d, %Y %I:%M:%S %p'`

In [ ]:
# Create a datetime object by passing the string values to the 'strptime()' function.
example1 = datetime.datetime.strptime('Dec 26, 2024 03:34:00', '%b %d, %Y %H:%M:%S')
print(example1)

In [ ]:
type(example1)

---

### Activity 10: Only `dates` & `times`

We can create separate `datetime.date` and `datetime.time` objects by passing values to the `date()` and `time()` functions.

In [ ]:
# Create only a 'datetime.date' object for today. Print it and its data-type.
today = datetime.date(2024, 7, 17)
print(type(today))
print(today)

In [ ]:
# Create only a 'datetime.time' object for any random time. Print it and its data-type.
exam_time = datetime.time(hour=14, minute=30, second=12)
print(type(exam_time))
print(exam_time)

In [ ]:
date_time

In [ ]:
air_df.head()

---

### Activity 11: Continuing Air Quality Analysis



Let's remove the `Date` & `Time` columns from the DataFrame because we don't need them and insert the `data_time` in it because it contains the `datetime` objects.

Let's add the Pandas series to the DataFrame at `index = 0`. Also, let's label it as a `DateTime` column.

In [ ]:
# Remove the Date & Time columns from the DataFrame and insert new series
air_df.drop(columns=["Date","Time"],axis=1,inplace=True)

In [ ]:
air_df.head()

In [ ]:
# Display the first five rows of the DataFrame after removing the 'Date' and 'Time' columns and adding the 'DateTime' column.

air_df.insert(0,"DateTime",date_time)

In [ ]:
air_df.head()

---

### Activity 12: Extract Year, Month, Day & Weekday Values

Let's add four more columns to the DataFrame. The should contain the year, month, day and day-name values for each observation on the air pollutants, temperature, relative humidity and absolute humidity.

For this you can apply the following attributes/functions:

- `series_name.dt.year` to get a Pandas series containing the year values as integers.

- `series_name.dt.month` to get a Pandas series containing the month values as integers.

- `series_name.dt.day` to get a Pandas series containing the day values as integers.

- `series_name.dt.day_name()` to get a Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.


In [ ]:
# Get the Pandas series containing the year values as integers.
year_series=date_time.dt.year
year_series

In [ ]:
# Get the Pandas series containing the month values as integers.
month_series=date_time.dt.month
month_series

In [ ]:
#Get the Pandas series containing the day values as integers.
day_series=date_time.dt.day
day_series

In [ ]:
# Get the Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.
day_nameseries=date_time.dt.day_name()
print(day_nameseries)

We can add a column to a DataFrame by following the syntax given below.

**Syntax:** `df_name['column_name'] = pandas_series`

where `df_name` is the Pandas DataFrame in which the `pandas_series` to added as a column with the `column_name` as the desired name for the column.

**Note:** The indices of the items contained in the `pandas_series` must be the same as the indices of the `df_name` DataFrame.

In [ ]:
# Add the 'Year', 'Month', 'Day' and 'Day Name' columns to the DataFrame.
air_df['Year'] = year_series
air_df['Month'] = month_series
air_df['Day'] = day_series
air_df['day_name'] = day_nameseries

In [ ]:
air_df.info()

In [ ]:
air_df.head()

Let's sort the DataFrame by the `DateTime` values in the ascending order by using the `sort_values()` function. Inside the function, you need to pass the `by = 'DateTime'` parameter to sort the DataFrame by the `DateTime` values.

**Note:** By default, the `sort_values` function sorts a DataFrame in the ascending order. To sort it in the descending order, pass `ascending=False` as the second parameter inside the `sort_values` function.

In [ ]:
#  Sort the DataFrame by the 'DateTime' values in the ascending order.
air_df = air_df.sort_values(by='DateTime')

In [ ]:
air_df